### PR TITLE
feat: [GH-177] forward all cache options to key value cache

### PR DESCRIPTION
**Pull request by HishamAli81**
To be able to support custom key value caches that have additional cache options, ensure to forward all cache options to the key value cache when setting a entry into the cache.

### CODE REVIEW

File: src/__tests__/HTTPCache.test.ts [hunk 1]
```diff
@@ -409,6 +410,32 @@ describe('HTTPCache', () => {
...
+    const storeSet = jest.spyOn(store, 'set');
+
+    const { cacheWritePromise } = await httpCache.fetch(
+      adaUrl,
+      {},
+      {
+        cacheOptions: {
+          ttl: 20,
+          tags: ['foo', 'bar'],
+        } as unknown as CacheOptions,
```
Comment (trevor-scheer) [inline src/__tests__/HTTPCache.test.ts:425]: @HishamAli81 The typings story here seems half-baked. It looks like this PR wants the `CacheOptions` type to be configurable, which would mean the `KeyValueCache.KeyValueCacheSetOptions` type would need to be configurable as well. 

So the required changes I think are necessary to get this to a good place are:
* `KeyValueCache`
  * `interface KeyValueCache<V = string, TCacheSetOptions extends KeyValueCacheSetOptions = KeyValueCacheSetOptions>`
  * (this creates more flexibility in what can be provided as opts to KVC and lets us drill the type down)
* Update this PR, require new version of `@apollo/utils.keyvaluecache` once it's published, and add a similar generic type to `HTTPCache` which is used throughout and drilled down to KVC.

Are you up for that? I know I'm taking you on a longer ride than you expected.

Comment (trevor-scheer) [reply src/__tests__/HTTPCache.test.ts:425]: This is where that package lives btw: https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache

Comment (HishamAli81) [reply src/__tests__/HTTPCache.test.ts:425]: @trevor-scheer Thanks for the feedback, and I totally agree with the direction you've suggested - I'll follow up.

Comment (trevor-scheer) [reply src/__tests__/HTTPCache.test.ts:425]: Hey @HishamAli81, just checking in. Would you like me to keep this PR open?

**Comment by trevor-scheer** [comment]
Closing as stale but still open to this change.

**Comment by HishamAli81** [comment]
> Closing as stale but still open to this change.

@trevor-scheer Apologies for the delay, just circling back to this. I'll close this PR and will create the apollo-utils PR like you suggested first, then will create a new PR for this repo.

**Comment by trevor-scheer** [comment]
@HishamAli81 no worries. I'm going to reopen this one in case you want to pick up here - that would be my preference. Feel free to close it if you'd rather just open a new one still.

**Comment by HishamAli81** [comment]
@trevor-scheer I've updated this PR with the latest version of `@apollo/utils.keyvaluecache` for better type support when customizing `CacheOptions`

File: src/__tests__/RESTDataSource.test.ts [hunk 2]
```diff
@@ -16,6 +16,10 @@ import type { WithRequired } from '@apollo/utils.withrequired';
 import { Headers as NodeFetchHeaders } from 'node-fetch';
 import type { Logger } from '@apollo/utils.logger';
 
+interface CustomCacheOptions extends CacheOptions {
+  tags?: string[];
+}
```
Comment (trevor-scheer) [inline src/__tests__/RESTDataSource.test.ts:21]: Maybe just bump this down inside the one test where it's used?

File: .changeset/plenty-cooks-bathe.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@apollo/datasource-rest': minor
+---
+
+Add support for custom cache set options
```
Comment (trevor-scheer) [inline .changeset/plenty-cooks-bathe.md:5]: Can you provide an example and use case here just to let users know why they might want this or what problem it's solving?

**Review by trevor-scheer — APPROVED** [review]
Couple minor changes but overall LGTM! Happy to get this released next week.

**Comment by HishamAli81** [comment]
@trevor-scheer Thanks for the review, I've just pushed out the minor updates you suggested.

**Comment by trevor-scheer** [comment]
@HishamAli81 Thanks for all of your efforts on this! Kicking off a release shortly.

**Pull request merged**